### PR TITLE
Fix libsodium hanging

### PR DESCRIPTION
**Pull request by defntvdm**
Should solve #632

### CODE REVIEW

File: internal/crypto/libsodium/reader.go [hunk 1]
```diff
@@ -22,40 +23,56 @@ type Reader struct {
 
 	outIdx int
 	outLen int
+
+	// In case of using io.Pipe we can't read header until writer doesn't write, therefor we use these sync
+	o         sync.Once
```
Comment (g0djan) [inline internal/crypto/libsodium/reader.go:28]: "o" rename it

File: internal/crypto/libsodium/writer.go [hunk 2]
```diff
@@ -21,36 +22,49 @@ type Writer struct {
 	out []byte
 
 	inIdx int
+
+	// In case of using io.Pipe we can't write header until reader doesn't read, therefor we use these sync
+	o         sync.Once
```
Comment (g0djan) [inline internal/crypto/libsodium/writer.go:27]: And this

**Comment by g0djan** [comment]
Hi, thanks a lot for this fix. Can you explain the root causes of hanging wal-push?
I understand that it could happen without crypter.mutex.RUnlock(). What did you fix in other files?

**Comment by defntvdm** [comment]
Hi! The root of problem is io.Pipe. I wrote comment above onceHeader in libsodium Reader and Writer. We started to write header in Writer creation, so it locked until Reader start read, but Reader will never start read, because it created after Writer. And the same problem in Reader.
RUnlock don't solve problem, it's just good practice to unlock everything we locked.

File: internal/crypto/libsodium/reader.go (hunk 1 above)
Comment (defntvdm) [reply internal/crypto/libsodium/reader.go]: renamed

File: internal/crypto/libsodium/writer.go (hunk 2 above)
Comment (defntvdm) [reply internal/crypto/libsodium/writer.go]: renamed

**Pull request merged**